### ISSUE TITLE
Performance issues when using oversuscription

### ISSUE THREAD

**Issue by dalcinl** [open-mpi/ompi#10426]
I've been testing mpi4py against `ompi/main` using an automatic schedule that runs on a weekly basis on GitHub Actions. 

The GitHub-hosted runners have only two virtual cores, but I run tests with up to 5 MPI processes turning on oversubscription (setting  `rmaps_default_mapping_policy = :oversubscribe` within `$HOME/.prte/mca-params.conf`).
The current workflow file is [here](https://github.com/mpi4py/mpi4py-testing/blob/master/.github/workflows/openmpi.yml).

Up to January 29, the testsuite used to take around 4m 30s to run to completion [[link]](https://github.com/mpi4py/mpi4py-testing/runs/4988029360).
Since February 5, the same testsuite  needs around 1h 20m to finish [[link]](https://github.com/mpi4py/mpi4py-testing/runs/5074388856).
Unfortunately, the actual logs are long gone, but at least you can see the elapsed time (look for the **Test mpi4py (np=5)** line).

Is this a regression or some expected change in behavior? 
Is there a new configuration option I'm missing that would allow me to go back to the previous behavior?

**Comment by jsquyres** [comment]
@awlauria Can you answer @dalcinl's question?  If there's a config change that is needed, this definitely needs to be documented for v5.0.x.

**Comment by devreal** [comment]
@dalcinl I don't remember any changes this year that could cause this but I had a patch a while back that added an option to control how MPI processes yield the CPU when idle. The default is still `sched_yield` but you can set `--mca threads_pthreads_yield_strategy nanosleep` to use the `nanosleep` call instead (and `--mca threads_pthreads_nanosleep_time X` to control the time slept; default is the minimum of a nanosecond, just to trigger the scheduler). This seems to do a better job at forcing processes to be rescheduled. Maybe that will help bring down your run times when oversubscribing. It shouldn't have any negative performance impact if not oversubscribing.

**Comment by dalcinl** [comment]
@devreal My main concern is that I made absolutely no change to the way I configure Open MPI (either compile time or runtime), and yet I get this huge performance regression. Perhaps it is related to some change in GitHub Actions and how the runners are setup or configured. In the mean time, I'll try your suggestion and report back on the outcome.

**Comment by dalcinl** [comment]
@devreal I tried your suggestion (I kept the default time), but [had no effect](https://github.com/mpi4py/mpi4py-testing/runs/6587359633?check_suite_focus=true)

**Comment by dalcinl** [comment]
@devreal @awlauria After manually bisecting the issue :sweat:, I can tell for sure the regression is within Open MPI.

The last good commit is 3b4d64f [[logs]](https://github.com/mpi4py/mpi4py-testing/runs/6591499931?check_suite_focus=true), running mpi4py testsuite on 5 MPI processes takes under 3 minutes.

The regression is introduced after merging #9097 (MPI Sessions PR) in commit 7291361 [[logs]](https://github.com/mpi4py/mpi4py-testing/runs/6591227291?check_suite_focus=true) (cancelled after 9 minutes of running  on 3 MPI processes).

@hppritcha From your recent interactions in other issues I submitted, I believe you may want to get involved in this one.

**Comment by devreal** [comment]
Thanks for checking. Let me see if I can reproduce that on my machine

**Comment by hppritcha** [comment]
looking in to this.

**Comment by hppritcha** [comment]
looks like a piece from 2b335ed45370c1ac29379654e099038b2f627ad7 disappeared with the sessions support

**Comment by hppritcha** [comment]
@dalcinl could you try the patch at #10428 ?

You should see a message like this:
```
turning on yieldy thingy
turning on yieldy thingy
turning on yieldy thingy
turning on yieldy thingy
turning on yieldy thingy
```
if its working for your oversubscription case.

**Comment by dalcinl** [comment]
@devreal Look like your suggestion was not enough to workaround the issue. I'm just trying to figure out why.
Did you forget to mention that I should have set the MCA parameter `mpi_yield_when_idle = true` as well? Perhaps you assumed that yield_when_idle was already true, given that I was oversubscribing, and you did not know about the bug @hppritcha is fixing in #10428. Is either that, or the nanonsleep-based yield misbehaves or is broken, or the default time is not enough. Am I missing something?

**Comment by devreal** [comment]
Yes, I was assuming that this was set automatically but it looks like it isn't. Have you tried setting `mpi_yield_when_idle = true` manually?

**Comment by hppritcha** [comment]
i think the intent of https://github.com/open-mpi/ompi/commit/2b335ed45370c1ac29379654e099038b2f627ad7 was to avoid forcing the user to set additional ompi mca parameters if prte knows that the ompi app is being run in an oversubscribed manner.

**Comment by dalcinl** [comment]
@hppritcha As I commented in #10428, your patch worked fine. My question to Joseph was unrelated to your fix, I just wanted to know why his suggestion did not work.

**Comment by awlauria** [comment]
Merged to main and v5.0.x - closing.
